openlibrary/fastapi/search.py

```diff
@@ -12,47 +17,101 @@
+    """
+
+    q: str = Query("", description="The search query, like keyword.")
+
+    # from check_params in works.py
+    title: str | None = None
+    publisher: str | None = None
+    oclc: str | None = None
+    lccn: str | None = None
+    contributor: str | None = None
+    subject: str | None = None
+    place: str | None = None
+    person: str | None = None
+    time: str | None = None
+    # from workscheme facet_fields
+    has_fulltext: bool | None = None
+    public_scan_b: bool | None = None
+
+    """
+    The day will come when someone asks, why do we have Field wrapping Query
+    The answer seems to be:
+    1. Depends(): Tells FastAPI to explode the Pydantic model into individual arguments (dependency injection).
+    2. Field(Query([])): Overrides the default behavior for lists. It forces FastAPI to look for ?author_key=...
+       in the URL query string instead of expecting a JSON array in the request body.
+    The Field part is needed because FastAPI's default guess for lists inside Pydantic models is wrong for this use case.
+       It guesses "JSON Body," and you have to manually correct it to "Query String."
+    See: https://github.com/internetarchive/openlibrary/pull/11517#issuecomment-3584196385
+    """
+    author_key: list[str] = Field(Query([]))
+    subject_facet: list[str] = Field(Query([]))
+    person_facet: list[str] = Field(Query([]))
+    place_facet: list[str] = Field(Query([]))
+    time_facet: list[str] = Field(Query([]))
+    first_publish_year: list[str] = Field(Query([]))
+    publisher_facet: list[str] = Field(Query([]))
+    language: list[str] = Field(Query([]))
+    author_facet: list[str] = Field(Query([]))
+
+
+@router.get("/search.json")
 async def search_json(
     request: Request,
-    q: str | None = Query("", description="The search query."),
-    page: int | None = Query(
-        1, ge=1, description="The page number of results to return."
-    ),
-    limit: int = Query(
-        100, ge=1, le=1000, description="The number of results per page."
-    ),
+    pagination: Annotated[Pagination, Depends()],
+    public_query_options: Annotated[PublicQueryOptions, Depends()],
     sort: str | None = Query(None, description="The sort order of results."),
-    offset: int | None = Query(None, description="The offset of results to return."),
     fields: str | None = Query(None, description="The fields to return."),
-    # facet: bool = Query(False, description="Whether to return facets."),
     spellcheck_count: int | None = Query(
-        3, description="The number of spellcheck suggestions."
+        default_spellcheck_count, description="The number of spellcheck suggestions."
+    ),
+    query_str: str | None = Query(
+        None, alias="query", description="A full JSON encoded solr query."
     ),
 ):
     """
     Performs a search for documents based on the provided query.
     """
+    query: dict[str, Any] = {}
+    if query_str:
+        query = json.loads(query_str)
```

Comment:
In the previous version, sort/page/etc were also in the `?query={...}` parameter

Reply:
If you mean the previous version of this PR then I think it was just a mistake.
If you mean the web.py version of this endpoint then I'm not sure where you see that can you elaborate?

They certainly can specify those things in the `?query={...}` param but we don't really check/validate that as of now. Would be good idea for the future though.